### ISSUE TITLE
Gradients do not exist for variables 'maxvit/block_00_00/attention/relative_bias:0'

### ISSUE THREAD

**Issue by leondgarse** [google-research/maxvit#7]
- In [maxvit.py class Attention #L216-L238](https://github.com/google-research/maxvit/blob/main/maxvit/models/maxvit.py#L216-L238) implementation, we have `self.relative_bias = self.add_weight(...)` added as weight first, then rearranged `self.reindexed_bias = attn_utils.reindex_2d_einsum_lookup(...)` in `build`, not in `call`. I'm not sure about this, but in my understanding, the second rearranged `self.reindexed_bias` is a copy of the first initialized weights, and thus the original weights will not get updated during training. I think the `self.reindexed_bias = attn_utils.reindex_2d_einsum_lookup(...)` operation should be moved into `def call` block, like into [maxvit.py#L258](https://github.com/google-research/maxvit/blob/df590d74b615d2f1a8e52f95490168254aae6443/maxvit/models/maxvit.py#L258).
- This [Colab maxvit_train_test.ipynb](https://colab.research.google.com/drive/1sA1KLP8boelxL7exH5Yp35R2vc9rdcwr?usp=sharing) is a basic training test of whether `relative_bias` is changed during training. Anyway, maybe it's I'm not using it correctly. please help confirm.
- BTW, this is my keras implementation of MaxViT [Github keras_cv_attention_models/maxvit](https://github.com/leondgarse/keras_cv_attention_models/tree/main/keras_cv_attention_models/maxvit). Currently It's only `MaxViT_T` weights ported, as the ported `relative_bias` seems not right...

**Comment by vztu** [comment]
Thanks @leondgarse for catching this. I guess you might be right. We'll investigate this issue on our side. Our whole training pipeline is in TF1.x, not sure whether it is about this issue.

With that being said, you would still be able to load our checkpoint and do inference normally? I mean even if the `self.relative_bias` is a randomly initialized weights that never get updated during training, you can just treat it as a constant bias and do whatever you want. I would guess it wouldn't hurt the performance that much, judging from our current results.

**Comment by leondgarse** [comment]
Yes, actually the ported model accuracy is pretty norml. My `MaxViT_T` model tested accuracy is `top1: 83.532 top5: 96.664`.

**Comment by vztu** [comment]
Thanks for confirming the results. Given that it works normally, we're closing this issue for now. Feel free to open another issue if you find something new.

**Comment by leondgarse** [comment]
It's actually not the accuracy bothering me. As you said, we can even regard it as a constant bias, but the potential of the model may reach a better accuracy, with a fully trainable positional embedding. Ya, it may not the case for TF 1. Anyway is OK for me. :)

**Comment by vztu** [comment]
Yeah we totally agree with you on this point. We'll look into it on our side and investigate this issue. However, I'll keep this code as it is since it is the original code to reproduce the paper's results. But anyway thanks for spotting this issue!

**Comment by leondgarse** [comment]
I think there's an issue here, for `384` and `512` model weights. As told in [Exact Training Details to Reproduce Tiny Model Results #8](https://github.com/google-research/maxvit/issues/8), that some of `384` / `512` models are fine-tuned from `224` ones. Thus in my understanding, there could be a process like:
- When fine-tuning on new resolution `384`, model `self.relative_bias` weights is firstly randomly initialized, then arranged to `self.reindexed_bias`.
- Then we reload weights from previous `224` one, the weights is restored to `self.relative_bias`, but not to `self.reindexed_bias`. They are totally different at this point, and the later one is used in actual training.
- After fine-tuning training is done, we save model weights. This time, it's the added weights `self.relative_bias` been saved, not the actual `self.reindexed_bias` used in training.
In a word, the saved `self.relative_bias` in weights is not the actually used one.
I took some tests using simillar process for `MaxViT_Tiny` `224` / `384` / `512` weights, but cannot restore `384` / `512` acuracy.
| Model       | resolution | central_crop | resize method | antialias | Acc         |
| ----------- | ---------- | ------------ | ------------- | --------- | ----------- |
| MaxViT_Tiny | 224        | 0.95         | bicubic       | True      | 0.83412     |
| MaxViT_Tiny | 224        | 0.95         | bicubic       | False     | 0.83532     |
| MaxViT_Tiny | 224        | 0.875        | bicubic       | False     | 0.83532     |
| MaxViT_Tiny | 224        | 0.95         | bilinear      | True      | **0.83644** |
| MaxViT_Tiny | 224        | 0.95         | bilinear      | False     | 0.83644     |
| MaxViT_Tiny | 224        | 0.875        | bilinear      | False     | 0.83524     |
| MaxViT_Tiny | 224        | 0.875        | bilinear      | True      | 0.83362     |
| MaxViT_Tiny | 224        | 0.99         | bilinear      | True      | 0.83282     |
|             |            |              |               |           |             |
| MaxViT_Tiny | 384        | 0.95         | bicubic       | True      | 0.83706     |
| MaxViT_Tiny | 384        | 0.95         | bilinear      | True      | **0.83738** |
| MaxViT_Tiny | 384        | 0.95         | bilinear      | False     | 0.8373      |
| MaxViT_Tiny | 384        | 0.95         | bicubic       | False     | 0.83478     |
| MaxViT_Tiny | 384        | 0.923        | bilinear      | False     | 0.83592     |
| MaxViT_Tiny | 384        | 0.8333       | bicubic       | True      | 0.8276      |
|             |            |              |               |           |             |
| MaxViT_Tiny | 512        | 0.95         | bilinear      | True      | **0.84222** |

**Comment by vztu** [comment]
I'm wondering what is the result if you test with:
* no central_crop, just directly resizing using `tf.image.resize` with default settings. Say at 384, just resize to 384.

**Comment by leondgarse** [comment]
Amazing! I actually followed [utils.py _decode_and_center_crop](https://github.com/google-research/maxvit/blob/main/maxvit/models/utils.py#L108) process in the first place. Ya, turning off `central_crop` imrpoved the accuracy... Also, `tf.image.resize` default settings are `method='bilinear', antialias=False`, I've tried some others:

| Model       | resolution | central_crop | resize method | antialias | Acc         |
| ----------- | ---------- | ------------ | ------------- | --------- | ----------- |
| MaxViT_Tiny | 224        | 0.95         | bilinear      | False     | 0.83644     |
| MaxViT_Tiny | 224        | 0            | bilinear      | False     | 0.83116     |
|             |            |              |               |           |             |
| MaxViT_Tiny | 384        | 0.95         | bilinear      | True      | 0.83738     |
| MaxViT_Tiny | 384        | 0            | bilinear      | True      | 0.8422      |
| MaxViT_Tiny | 384        | 0            | bilinear      | False     | **0.84304** |
| MaxViT_Tiny | 384        | 0            | bicubic       | False     | 0.84182     |
| MaxViT_Tiny | 384        | 0            | bicubic       | True      | 0.8418      |
|             |            |              |               |           |             |
| MaxViT_Tiny | 512        | 0.95         | bilinear      | True      | 0.84222     |
| MaxViT_Tiny | 512        | 0            | bilinear      | True      | 0.84486     |
| MaxViT_Tiny | 512        | 0            | bilinear      | False     | 0.8487      |
| MaxViT_Tiny | 512        | 0            | bicubic       | False     | **0.84902** |
| MaxViT_Tiny | 512        | 0            | bicubic       | True      | 0.84894     |
